src/driver-marsgaming/marsgaming-buttons.c

```diff
@@ -0,0 +1,425 @@
+			case KEY_LEFTMETA: modifiers |= MODIFIER_LEFTMETA; break;
+			case KEY_RIGHTCTRL: modifiers |= MODIFIER_RIGHTCTRL; break;
+			case KEY_RIGHTSHIFT: modifiers |= MODIFIER_RIGHTSHIFT; break;
+			case KEY_RIGHTALT: modifiers |= MODIFIER_RIGHTALT; break;
+			case KEY_RIGHTMETA: modifiers |= MODIFIER_RIGHTMETA; break;
+			default:
+				if (key0 == KEY_RESERVED) {
+					key0 = event.event.key;
+					++keys_pressed;
+				} else if (key1 == KEY_RESERVED) {
+					key1 = event.event.key;
+					++keys_pressed;
+				} else {
+					return -EINVAL;
+				}
+			}
+			break;
+		case RATBAG_MACRO_EVENT_KEY_RELEASED:
+			switch (event.event.key) {
+			case KEY_LEFTCTRL: modifiers &= ~MODIFIER_LEFTCTRL; break;
+			case KEY_LEFTSHIFT: modifiers &= ~MODIFIER_LEFTSHIFT; break;
+			case KEY_LEFTALT: modifiers &= ~MODIFIER_LEFTALT; break;
+			case KEY_LEFTMETA: modifiers &= ~MODIFIER_LEFTMETA; break;
+			case KEY_RIGHTCTRL: modifiers &= ~MODIFIER_RIGHTCTRL; break;
+			case KEY_RIGHTSHIFT: modifiers &= ~MODIFIER_RIGHTSHIFT; break;
+			case KEY_RIGHTALT: modifiers &= ~MODIFIER_RIGHTALT; break;
+			case KEY_RIGHTMETA: modifiers &= ~MODIFIER_RIGHTMETA; break;
+			default:
+				// As soon as we release a key that we pressed, return what we have processed so far
+				if (event.event.key == key0 || event.event.key == key1) {
+					*key0_out = key0;
+					*key1_out = key1;
+					*modifiers_out = modifiers;
+					return 2;
+				}
+
+				return -EINVAL;
+			}
+		case RATBAG_MACRO_EVENT_WAIT:
+			break;
+		default:
+			return -EINVAL;
+		}
+	}
+
+	return -EINVAL;
+}
+
+struct marsgaming_button_action_to_ratbag_parser {
+	enum marsgaming_button_action marsgaming_action_id;
+	struct ratbag_button_action (*parse_action)(struct ratbag_button *button,
+						    const struct marsgaming_button_info *button_info);
+};
+
+static const struct marsgaming_button_action_to_ratbag_parser marsgaming_button_action_to_ratbag_parsers[] = {
+	{ MARSGAMING_MM4_ACTION_LEFT_CLICK, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_RIGHT_CLICK, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_MIDDLE_CLICK, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_BACKWARD, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_FORWARD, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_DPI_SWITCH, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_DPI_MINUS, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_DPI_PLUS, marsgaming_button_action_lookup },
+	{ MARSGAMING_MM4_ACTION_PROFILE_SWITCH, marsgaming_button_action_lookup },
+	//{ MARSGAMING_MM4_ACTION_MEDIA, marsgaming_button_action_media },
+	{ MARSGAMING_MM4_ACTION_COMBO_KEY, marsgaming_button_action_key },
+	{ MARSGAMING_MM4_ACTION_SINGLE_KEY, marsgaming_button_action_key },
+	//{ MARSGAMING_MM4_ACTION_MACRO, marsgaming_button_action_macro }, // TODO: Implement macros
+	//{ MARSGAMING_MM4_ACTION_FIRE, marsgaming_button_action_fire },
```

Comment:
There are no `marsgaming_button_action_media`, `marsgaming_button_action_macro`  and `marsgaming_button_action_fire`. Maybe stub them out and add TODOs in there with short description of what it should be?

They could look like this for now:

```c
static struct ratbag_button_action
marsgaming_button_action_media(struct ratbag_button *button,
			     const struct marsgaming_button_info *button_info)
{
	// TODO: something-something description of what should be done here.
	return (struct ratbag_button_action)BUTTON_ACTION_UNKNOWN;
}
```